gc.c

```diff
@@ -2785,13 +2785,49 @@ rb_objspace_data_type_name(VALUE obj)
     }
 }
 
+static int
+ptr_in_page_body_p(const void *ptr, const void *memb)
+{
+    struct heap_page *hpage = (struct heap_page *)(*(uintptr_t *)memb);
+    uintptr_t p_body = (uintptr_t)GET_PAGE_BODY(hpage->start);
+
+    if ((uintptr_t)ptr >= p_body) {
+        return (uintptr_t)ptr < (p_body + HEAP_PAGE_SIZE) ? 0 : 1;
```

Comment:
Peter and I resolved this offline but I'm documenting it here for posterity. 

I chose to extract this function that searches the whole heap page body instead of just the slots deliberately because it enables some follow on work that I want to do that relies on this functionality